
project build/make/
diff --git a/tools/releasetools/common.py b/tools/releasetools/common.py
index 71679e8cb2..72fd4d3780 100644
--- a/tools/releasetools/common.py
+++ b/tools/releasetools/common.py
@@ -766,6 +766,18 @@ def GetBootableImage(name, prebuilt_name, unpack_dir, tree_subdir,
   the source files in 'unpack_dir'/'tree_subdir'."""
 
   prebuilt_path = os.path.join(unpack_dir, "BOOTABLE_IMAGES", prebuilt_name)
+
+  # EXTENDROM: Magisk patcher
+  if (prebuilt_name == "boot.img" or prebuilt_name == "init_boot.img") and os.path.exists(prebuilt_path):
+    if info_dict is None:
+        info_dict = OPTIONS.info_dict
+    img = tempfile.NamedTemporaryFile()
+    shutil.copy(prebuilt_path, img.name)
+    logger.info("Patching (GetBootableImage): %s", img.name)
+    mpatcher = [ 'vendor/extendrom/config/magisk/patch.sh', img.name ]
+    RunAndCheckOutput(mpatcher, verbose=True, env={'system_root_image': info_dict.get("system_root_image", "false")})
+    shutil.copy(img.name, prebuilt_path)
+
   if os.path.exists(prebuilt_path):
     logger.info("using prebuilt %s from BOOTABLE_IMAGES...", prebuilt_name)
     return File.FromLocalFile(name, prebuilt_path)
