
project build/make/
diff --git a/tools/releasetools/sign_target_files_apks.py b/tools/releasetools/sign_target_files_apks.py
index d86014a01..c6ae5b0f6 100755
--- a/tools/releasetools/sign_target_files_apks.py
+++ b/tools/releasetools/sign_target_files_apks.py
@@ -737,13 +737,14 @@ def ReplaceOtaKeys(input_tf_zip, output_tf_zip, misc_info):
     extra_recovery_keys = []

   mapped_keys = []
-  for k in keylist:
-    m = re.match(r"^(.*)\.x509\.pem$", k)
+  for kl in keylist:
+    k = kl.decode('ISO-8859-1')
+    m = re.match(r"^(.*)$", k)
     if not m:
       raise common.ExternalError(
           "can't parse \"%s\" from META/otakeys.txt" % (k,))
     k = m.group(1)
-    mapped_keys.append(OPTIONS.key_map.get(k, k) + ".x509.pem")
+    mapped_keys.append(OPTIONS.key_map.get(k, k))

   if mapped_keys:
     print("using:\n   ", "\n   ".join(mapped_keys))
