
project frameworks/base/
diff --git a/services/core/java/com/android/server/pm/ComputerEngine.java b/services/core/java/com/android/server/pm/ComputerEngine.java
index 6047737af2ab..2c5997312d7c 100644
--- a/services/core/java/com/android/server/pm/ComputerEngine.java
+++ b/services/core/java/com/android/server/pm/ComputerEngine.java
@@ -1648,9 +1648,10 @@ public class ComputerEngine implements Computer {
                         state.getFirstInstallTime(), ps.getLastUpdateTime(), permissions, state,
                         userId, ps), permissions);
             } else {
-                packageInfo = PackageInfoUtils.generate(p, gids, flags,
-                        state.getFirstInstallTime(), ps.getLastUpdateTime(), permissions, state,
-                        userId, ps);
+		// extendrom: signature spoofing patch
+		PackageInfo packageInfo = mayFakeSignature(p, PackageParser.generatePackageInfo(p, gids, flags,
+			state.getFirstInstallTime(), ps.getLastUpdateTime(), permissions, state, userId, ps),
+			permissions);
             }
 
             if (packageInfo == null) {
