
project frameworks/base/
diff --git a/core/java/android/app/Activity.java b/core/java/android/app/Activity.java
index 750e3d6f6fed..203baa862e3d 100644
--- a/core/java/android/app/Activity.java
+++ b/core/java/android/app/Activity.java
@@ -163,7 +163,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.concurrent.Executor;
 import java.util.function.Consumer;
-
+import android.provider.Settings.Secure;       // extendrom: signature spoofing patch
 
 /**
  * An activity is a single, focused thing that the user can do.  Almost all
@@ -5408,6 +5409,26 @@ public class Activity extends ContextThemeWrapper
             }
         }
 
+       /*
+       signature spoofing patch:
+       Only when the requested permission asks for signature spoofing, check if that is allowed
+       in developer options and skip the dialog if not
+       */
+       List<String> permlist = Arrays.asList(permissions);
+       if (permlist.contains("android.permission.FAKE_PACKAGE_SIGNATURE")) {
+           if (android.provider.Settings.Secure.getInt(getContentResolver(),
+                  android.provider.Settings.Secure.ALLOW_SIGNATURE_FAKE, 0) == 0) {
+               Log.w(TAG, "Requested signature spoofing permission ("
+                       + Arrays.toString(permissions)
+                       + ") has been denied as it is not enabled in developer options");
+               return;
+           } else {
+               Log.w(TAG, "The app's permission request for: "
+                       + Arrays.toString(permissions)
+                       + " will be processed as signature spoofing is enabled in developer options!");
+           }
+       }
+
         final Intent intent = getPackageManager().buildRequestPermissionsIntent(permissions);
         startActivityForResult(REQUEST_PERMISSIONS_WHO_PREFIX, intent, requestCode, null);
         mHasCurrentPermissionsRequest = true;
