
project packages/apps/Settings/
diff --git a/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java b/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
index 2834bf7ee76..13fb073c6e8 100644
--- a/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
+++ b/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
@@ -581,6 +581,8 @@ public class DevelopmentSettingsDashboardFragment extends RestrictedDashboardFra
                 bluetoothA2dpConfigStore, fragment));
         controllers.add(new SharedDataPreferenceController(context));
         controllers.add(new OverlaySettingsPreferenceController(context));
+        controllers.add(new SpoofSignaturePreferenceController(context));      // extendrom: signature spoofing patch
+        controllers.add(new SpoofSignatureInfo(context));       // extendrom: signature spoofing patch
 
         return controllers;
     }
