
project bootable/recovery/
diff --git a/otautil/verifier.cpp b/otautil/verifier.cpp
index 8a65566e..37f3bc88 100644
--- a/otautil/verifier.cpp
+++ b/otautil/verifier.cpp
@@ -196,6 +196,7 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
 
   bool need_sha1 = false;
   bool need_sha256 = false;
+  bool need_sha512 = false;
   for (const auto& key : keys) {
     switch (key.hash_len) {
       case SHA_DIGEST_LENGTH:
@@ -204,13 +205,19 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
       case SHA256_DIGEST_LENGTH:
         need_sha256 = true;
         break;
+      case SHA512_DIGEST_LENGTH:
+       need_sha512 = true;
+        break;
     }
   }
 
   SHA_CTX sha1_ctx;
   SHA256_CTX sha256_ctx;
+  SHA512_CTX sha512_ctx;
+
   SHA1_Init(&sha1_ctx);
   SHA256_Init(&sha256_ctx);
+  SHA512_Init(&sha512_ctx);
 
   std::vector<HasherUpdateCallback> hashers;
   if (need_sha1) {
@@ -221,6 +228,10 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
     hashers.emplace_back(
         std::bind(&SHA256_Update, &sha256_ctx, std::placeholders::_1, std::placeholders::_2));
   }
+  if (need_sha512) {
+    hashers.emplace_back(
+        std::bind(&SHA512_Update, &sha512_ctx, std::placeholders::_1, std::placeholders::_2));
+  }
 
   double frac = -1.0;
   uint64_t so_far = 0;
@@ -242,6 +253,8 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
   SHA1_Final(sha1, &sha1_ctx);
   uint8_t sha256[SHA256_DIGEST_LENGTH];
   SHA256_Final(sha256, &sha256_ctx);
+  uint8_t sha512[SHA512_DIGEST_LENGTH];
+  SHA512_Final(sha512, &sha512_ctx);
 
   const uint8_t* signature = eocd + eocd_size - signature_start;
   size_t signature_size = signature_start - FOOTER_SIZE;
@@ -257,8 +270,8 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
 
   // Check to make sure at least one of the keys matches the signature. Since any key can match,
   // we need to try each before determining a verification failure has happened.
-  for (size_t i = 0; i < keys.size(); i++) {
-    const auto& key = keys[i];
+  size_t i = 0;
+  for (const auto& key : keys) {
     const uint8_t* hash;
     int hash_nid;
     switch (key.hash_len) {
@@ -270,6 +283,10 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
         hash = sha256;
         hash_nid = NID_sha256;
         break;
+      case SHA512_DIGEST_LENGTH:
+        hash = sha512;
+        hash_nid = NID_sha512;
+        break;
       default:
         continue;
     }
@@ -291,11 +308,20 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
         continue;
       }
 
+      LOG(INFO) << "whole-file signature verified against EC key " << i;
+      return VERIFY_SUCCESS;
+    } else if (key.key_type == Certificate::KEY_TYPE_EC && key.hash_len == SHA512_DIGEST_LENGTH) {
+      if (!ECDSA_verify(0, hash, key.hash_len, sig_der.data(), sig_der.size(), key.ec.get())) {
+        LOG(INFO) << "failed to verify against EC key " << i;
+        continue;
+      }
+
       LOG(INFO) << "whole-file signature verified against EC key " << i;
       return VERIFY_SUCCESS;
     } else {
       LOG(INFO) << "Unknown key type " << key.key_type;
     }
+    i++;
   }
 
   if (need_sha1) {
@@ -304,6 +330,9 @@ int verify_file(VerifierInterface* package, const std::vector<Certificate>& keys
   if (need_sha256) {
     LOG(INFO) << "SHA-256 digest: " << print_hex(sha256, SHA256_DIGEST_LENGTH);
   }
+  if (need_sha512) {
+    LOG(INFO) << "SHA-512 digest: " << print_hex(sha512, SHA512_DIGEST_LENGTH);
+  }
   LOG(ERROR) << "failed to verify whole-file signature";
   return VERIFY_FAILURE;
 }
@@ -375,8 +404,8 @@ bool CheckRSAKey(const std::unique_ptr<RSA, RSADeleter>& rsa) {
   const BIGNUM* out_e;
   RSA_get0_key(rsa.get(), &out_n, &out_e, nullptr /* private exponent */);
   auto modulus_bits = BN_num_bits(out_n);
-  if (modulus_bits != 2048 && modulus_bits != 4096) {
-    LOG(ERROR) << "Modulus should be 2048 or 4096 bits long, actual: " << modulus_bits;
+  if (modulus_bits != 2048 && modulus_bits != 4096 && modulus_bits != 8192) {
+    LOG(ERROR) << "Modulus should be 2048, 4096 or 8192 bits long, actual: " << modulus_bits;
     return false;
   }
 
@@ -400,8 +429,8 @@ bool CheckECKey(const std::unique_ptr<EC_KEY, ECKEYDeleter>& ec_key) {
     return false;
   }
   auto degree = EC_GROUP_get_degree(ec_group);
-  if (degree != 256) {
-    LOG(ERROR) << "Field size of the ec key should be 256 bits long, actual: " << degree;
+  if (degree != 256 && degree != 512) {
+    LOG(ERROR) << "Field size of the ec key should be 256 or 512 bits long, actual: " << degree;
     return false;
   }
 
@@ -429,6 +458,9 @@ bool LoadCertificateFromBuffer(const std::vector<uint8_t>& pem_content, Certific
     case NID_sha1WithRSAEncryption:
       cert->hash_len = SHA_DIGEST_LENGTH;
       break;
+    case NID_sha512WithRSAEncryption:
+      cert->hash_len = SHA512_DIGEST_LENGTH;
+      break;
     case NID_sha256WithRSAEncryption:
     case NID_ecdsa_with_SHA256:
       cert->hash_len = SHA256_DIGEST_LENGTH;
