
project frameworks/base/
diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 2b4836861219..4c8674e0c482 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -8084,6 +8084,15 @@ public final class Settings {
         @Readable
         public static final String INSTALL_NON_MARKET_APPS = "install_non_market_apps";
 
+       /**
+         * extendrom: whether applications can fake a signature.
+         *
+         * <p>1 = permit apps to fake signature
+         * <p>0 = disable this feature
+         * @hide
+         */
+        public static final String ALLOW_SIGNATURE_FAKE = "allow_signature_fake";
+
         /**
          * A flag to tell {@link com.android.server.devicepolicy.DevicePolicyManagerService} that
          * the default for {@link #INSTALL_NON_MARKET_APPS} is reversed for this user on OTA. So it